compiler/src/dotty/tools/dotc/ast/Desugar.scala

```diff
@@ -410,17 +429,26 @@ object desugar {
     // For all other classes, the parent is AnyRef.
     val companions =
       if (isCaseClass) {
+        // The return type of the `apply` method
+        val applyResultTpt =
+          if (isEnumCase)
+            if (parents.isEmpty) enumClassTypeRef
+            else parents.reduceLeft(AndTypeTree)
+          else TypeTree()
+
         val parent =
           if (constrTparams.nonEmpty ||
               constrVparamss.length > 1 ||
               mods.is(Abstract) ||
               constr.mods.is(Private)) anyRef
+          else
             // todo: also use anyRef if constructor has a dependent method type (or rule that out)!
-          else (constrVparamss :\ classTypeRef) ((vparams, restpe) => Function(vparams map (_.tpt), restpe))
+            (constrVparamss :\ (if (isEnumCase) applyResultTpt else classTypeRef)) (
+              (vparams, restpe) => Function(vparams map (_.tpt), restpe))
```

Comment:
I may miss some point here, it seems to me the fold is useless here, because `constrVparamss` can only be of size 0?